python2/httplib2/cookie.py

```diff
@@ -0,0 +1,402 @@
+            self.save(self.filename)
+
+    def get(self, domain=None, path=None, name=None):
+        """Get cookies filtered by domain, path or name
+
+        If one parameter is specified, the preceding parameter(s) must also
+        be specified.
+
+        If name is specified, only returns the cookie with that name.
+        Otherwise, return a cookie list.
+
+        Return KeyError if no matching cookie found.
+        """
+        if name is not None:
+            assert domain and path, "domain and path must be given to get a cookie by name"
+            path = Cookie.normalize_path(path)
+            return self._cookies[domain][path][name]
+
+        if path is not None:
+            assert domain, "domain must be given to get cookies by path"
+            path = Cookie.normalize_path(path)
+            path_cookies = self._cookies[domain][path]
+            return list(path_cookies.values())
+
+        if domain is not None:
+            return [ck for pc in self._cookies[domain].itervalues() for ck in pc.itervalues()]
+
+        return [ck for dc in self._cookies.itervalues() for pc in dc.itervalues() for ck in pc.itervalues()]
+
+    def iter(self, domain=None, path=None):
+        """Return an iterator of cookies filtered by domain, path.
+
+        If domain and path are not specified, all cookies are returned.
+
+        Raise KeyError if no matching cookie found.
+        """
+        if path is not None:
+            assert domain, "domain must be given to get cookies by path"
+            path = Cookie.normalize_path(path)
+            for ck in self._cookies[domain][path].itervalues():
+                yield ck
+            return
+
+        if domain is not None:
+            for pc in self._cookies[domain].itervalues():
+                for ck in pc.itervalues():
+                    yield ck
+            return
+
+        for dc in self._cookies.itervalues():
+            for pc in dc.itervalues():
+                for ck in pc.itervalues():
+                    yield ck
+
+    def clear(self, domain=None, path=None, name=None):
+        """Delete cookies by domain, path, or name.
+
+        If one parameter is specified, the preceding parameter(s) must also
+        be specified.
+
+        No error if no matching cookie found.
+        """
+        try:
+            if name is not None:
+                assert domain and path, "domain and path must be given to remove a cookie by name"
+                path = Cookie.normalize_path(path)
+                del self._cookies[domain][path][name]
+                return
+
+            if path is not None:
+                assert domain, "domain must be given to remove cookies by path"
+                path = Cookie.normalize_path(path)
+                del self._cookies[domain][path]
+                return
+
+            if domain is not None:
+                del self._cookies[domain]
+                return
+        except KeyError:
+            return
```

Comment:
(important)

We should only skip exceptions around **minimal** code. Future version of `normalize_path` may raise KeyError that we must not hide.
```suggestion
        if name is not None:
            assert domain and path, "domain and path must be given to remove a cookie by name"
            path = Cookie.normalize_path(path)
            try:
                del self._cookies[domain][path][name]
            except KeyError:
                pass
            return

        if path is not None:
            assert domain, "domain must be given to remove cookies by path"
            path = Cookie.normalize_path(path)
            try:
                del self._cookies[domain][path]
            except KeyError:
                pass
            return

        if domain is not None:
            self._cookies.pop(domain, None)
            return
```

another option:
```
    empty = {}
    self._cookies.get(domain, empty).get(path, empty).pop(name, None)
```

Reply:
`self._cookies.pop(domain, None)` doesn't faster than `del self._cookies[domain]`.  Since we don't need the returned value here.

Reply:
Of course it is faster because creating exceptions is very expensive.

But maybe just as important, `pop` is shorter and that helps to read code faster.

```
In [1]: def f_del(d, k):
   ...:     try:
   ...:         del d[k]
   ...:     except KeyError:
   ...:         pass
   ...:

In [2]: def f_pop(d, k):
   ...:     d.pop(k, None)
   ...:

In [3]: %timeit f_del({}, 'key')
674 ns ± 34.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit f_pop({}, 'key')
339 ns ± 11.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
I think we should mostly consider the normal case. 
```
def f_del(d, k):
    try:
        del d[k]
    except:
        pass

def f_pop(d, k):
    d.pop(k, None)
    
%timeit f_del({'key': 1}, 'key')
240 ns ± 0.523 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
%timeit f_pop({'key': 1}, 'key')
337 ns ± 0.634 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
As a trick, it's shorter but may not help to read, may make people confusion. Because pop() means `delete` and `return`.  Here we don't need the return value.